packages/library/src/decode-transaction.ts

```diff
@@ -0,0 +1,75 @@
+import {
+    assertAccountsDecoded,
+    assertAccountsExist,
+    type FetchAccountsConfig,
+    fetchJsonParsedAccounts,
+} from '@solana/accounts';
+import { Address } from '@solana/addresses';
+import type { GetMultipleAccountsApi } from '@solana/rpc-core';
+import type { Rpc } from '@solana/rpc-transport';
+import {
+    type AddressesByLookupTableAddress,
+    type CompilableTransaction,
+    decompileTransaction,
+    getCompiledTransactionDecoder,
+    type ITransactionWithSignatures,
+} from '@solana/transactions';
+
+let compiledTransactionDecoder: ReturnType<typeof getCompiledTransactionDecoder> | undefined = undefined;
+
+type FetchedAddressLookup = {
+    addresses: Address[];
+};
+
+async function fetchLookupTables(
+    lookupTableAddresses: Address[],
+    rpc: Rpc<GetMultipleAccountsApi>,
+    config?: FetchAccountsConfig,
+): Promise<AddressesByLookupTableAddress> {
+    const fetchedLookupTables = await fetchJsonParsedAccounts<FetchedAddressLookup[]>(
+        rpc,
+        lookupTableAddresses,
+        config,
+    );
+    assertAccountsDecoded(fetchedLookupTables);
+    assertAccountsExist(fetchedLookupTables);
+
+    return fetchedLookupTables.reduce<AddressesByLookupTableAddress>((acc, lookup) => {
+        return {
+            ...acc,
+            [lookup.address]: lookup.data.addresses,
+        };
+    }, {});
+}
+
+type DecodeTransactionConfig = FetchAccountsConfig & {
+    lastValidBlockHeight?: bigint;
+};
+
+export async function decodeTransaction(
+    encodedTransaction: Uint8Array,
+    rpc: Rpc<GetMultipleAccountsApi>,
+    config?: DecodeTransactionConfig,
+): Promise<CompilableTransaction | (CompilableTransaction & ITransactionWithSignatures)> {
+    const { lastValidBlockHeight, ...fetchAccountsConfig } = config ?? {};
+
+    if (!compiledTransactionDecoder) compiledTransactionDecoder = getCompiledTransactionDecoder();
+    const compiledTransaction = compiledTransactionDecoder.decode(encodedTransaction);
+    const { compiledMessage } = compiledTransaction;
+
+    const lookupTables =
+        'addressTableLookups' in compiledMessage &&
+        compiledMessage.addressTableLookups !== undefined &&
+        compiledMessage.addressTableLookups.length > 0
+            ? compiledMessage.addressTableLookups
+            : [];
+    const lookupTableAddresses = lookupTables.map(l => l.lookupTableAddress);
+
+    const fetchedLookupTables =
+        lookupTableAddresses.length > 0 ? await fetchLookupTables(lookupTableAddresses, rpc, fetchAccountsConfig) : {};
+
+    return decompileTransaction(compiledTransaction, {
```

Comment:
I was thinking of pushing slightly more logic to the lower-lever `decodeTransaction` function so that its high-level version would literally just be: a) fetch LUTs b) delegate to lower-level function.

In other words, I think we should add the fetched LUT array as an argument of the lower-level function (so we no longer throw an error when a transaction uses LUTs) but it's still synchronous and doesn't add a dependency circle.

Reply:
I did it this way so that the lower level function is less opinionated and only takes the input that it needs to care about, which is just the addresses per lookup table.

A bit speculative but that would allow fetching the addresses from somewhere else other than an RPC call if you want, like a database or cache that doesn't store LUT objects. My guess is that a use case like Helius, where you're providing ~real time parsed transactions, wouldn't be possible without caching LUTs.

With this implementation you have a lot of flexibility. If you cache LUTs as-is then you can provide an `Rpc<GetMultipleAccountsApi>` implementation that returns values from your cache and use this higher-level function. But if you want to store something more optimised then you have that option too using the lower-level function, as long as you can get a map of addresses per LUT from somewhere.

Reply:
Oh sorry I was being stupid. I checked the low-level function on the wrong branch for comparison and missed all the nice changes from the previous branches of the stack. Ignore me. 😬